ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/transactions/layered/AbstractTransactionsLayer.java

```diff
@@ -0,0 +1,593 @@
+    final var currLayerTx = pendingTransactions.get(transactionHash);
+    if (currLayerTx == null) {
+      return nextLayer.getByHash(transactionHash);
+    }
+    return Optional.of(currLayerTx.getTransaction());
+  }
+
+  @Override
+  public boolean contains(final Transaction transaction) {
+    return pendingTransactions.containsKey(transaction.getHash())
+        || nextLayer.contains(transaction);
+  }
+
+  @Override
+  public List<PendingTransaction> getAll() {
+    final List<PendingTransaction> allNextLayers = nextLayer.getAll();
+    final List<PendingTransaction> allTxs =
+        new ArrayList<>(pendingTransactions.size() + allNextLayers.size());
+    allTxs.addAll(pendingTransactions.values());
+    allTxs.addAll(allNextLayers);
+    return allTxs;
+  }
+
+  @Override
+  public long getCumulativeUsedSpace() {
+    return getLayerSpaceUsed() + nextLayer.getCumulativeUsedSpace();
+  }
+
+  protected long getLayerSpaceUsed() {
+    return spaceUsed;
+  }
+
+  protected abstract TransactionAddedResult canAdd(
+      final PendingTransaction pendingTransaction, final int gap);
+
+  @Override
+  public TransactionAddedResult add(final PendingTransaction pendingTransaction, final int gap) {
+
+    // is replacing an existing one?
+    TransactionAddedResult addStatus = maybeReplaceTransaction(pendingTransaction);
+    if (addStatus == null) {
+      addStatus = canAdd(pendingTransaction, gap);
+    }
+
+    if (addStatus.equals(TRY_NEXT_LAYER)) {
+      return addToNextLayer(pendingTransaction, gap);
+    }
+
+    if (addStatus.isSuccess()) {
+      processAdded(pendingTransaction);
+      addStatus.maybeReplacedTransaction().ifPresent(this::replaced);
+
+      nextLayer.notifyAdded(pendingTransaction);
+
+      if (!maybeFull()) {
+        // if there is space try to see if the added tx filled some gaps
+        tryFillGap(addStatus, pendingTransaction);
+      }
+
+      notifyTransactionAdded(pendingTransaction);
+    } else {
+      final var rejectReason = addStatus.maybeInvalidReason().orElseThrow();
+      metrics.incrementRejected(false, rejectReason, name());
+      LOG.atTrace()
+          .setMessage("Transaction {} rejected reason {}")
+          .addArgument(pendingTransaction::toTraceLog)
+          .addArgument(rejectReason)
+          .log();
+    }
+
+    return addStatus;
+  }
+
+  private boolean maybeFull() {
+    final long cacheFreeSpace = cacheFreeSpace();
+    final int overflowTxsCount = pendingTransactions.size() - maxTransactionsNumber();
+    if (cacheFreeSpace < 0 || overflowTxsCount > 0) {
+      LOG.atDebug()
+          .setMessage("Layer full: {}")
+          .addArgument(
+              () ->
+                  cacheFreeSpace < 0
+                      ? "need to free " + (-cacheFreeSpace) + " space"
+                      : "need to evict " + overflowTxsCount + " transaction(s)")
+          .log();
+
+      evict(-cacheFreeSpace, overflowTxsCount);
```

Comment:
It looks like the flow for adding txs to a layer is "add and then maybe evict". Wondering what the reasoning is for this...why not check if the tx will exceed the layer limit _before_ adding it?

I think the Prioritised layer uses `canAdd` to achieve this instead of relying on evict after the fact. 

This seems to be an inconsistency in the approach depending on which layer you are in, but maybe I'm missing something.

Reply:
the strategy is always "add then evict if full" in every layer, if `canAdd` is true it means that it must be in this layer so we add it, then we check if something needs to be evicted. So `canAdd` does not need to take in consideration any limit.